src/proxy.c

```diff
@@ -216,60 +224,89 @@ static gboolean
 cc_proxy_msg_read (GIOChannel *source, GIOCondition condition,
 	struct watcher_proxy_data *proxy_data)
 {
-	GIOStatus status;
-	g_autofree gchar *buffer = NULL;
-	gsize bytes_read = 0;
-	gsize len = 0;
-	GError *error = NULL;
-	struct proxy_msg msg = { 0 };
-
-	if (condition == G_IO_HUP) {
-		g_io_channel_unref(source);
-		goto out;
-	}
-
-	/* read length of message data that follows */
-	do {
-		status = g_io_channel_read_chars (source,
-				(gchar *)&msg,
-				sizeof (msg.length) +
-				sizeof (msg.reserved),
-				&bytes_read,
-				&error);
-	} while (status == G_IO_STATUS_AGAIN);
-
-	if (status != G_IO_STATUS_NORMAL) {
-		goto out;
+	struct proxy_msg p_msg = { 0 };
+	gchar iov_buffer[LINE_MAX] = { 0 };
+	ssize_t bytes_read;
+	gboolean fd_read = false;
+	struct msghdr msg = { 0 };
+	struct cmsghdr *cmsg = NULL;
+	struct iovec io = { .iov_base = iov_buffer,
+	                    .iov_len = sizeof(iov_buffer) };
+	char ctl_buffer[LINE_MAX] = { 0 };
+	unsigned char *data;
+	/* length + reserved */
+	const size_t data_offset = sizeof(p_msg.length) + sizeof(p_msg.reserved);
+	char buf_len[4] = { 0 };
```

Comment:
Can you add a comment explaining that this is sized to fit a 32-bit int.

It would be clearer and safer to do something like:

```
char buf_len[__SIZEOF_INT__] = { 0 };
```